##### Ahmed Bilal Azhar
##### 23I-2581


##### Setup

In [1]:
import pandas as pd
import numpy as np

##### Load Dataset and Initial Inspection

In [2]:
df = pd.read_csv('adult.csv')
df.head()

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,?,77053,HS-grad,9,Widowed,?,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,?,186061,Some-college,10,Widowed,?,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [4]:
df.shape

(32561, 15)

In [5]:
df.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education.num',
       'marital.status', 'occupation', 'relationship', 'race', 'sex',
       'capital.gain', 'capital.loss', 'hours.per.week', 'native.country',
       'income'],
      dtype='object')

In [6]:
df.sample(5)

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
15258,21,Private,263641,HS-grad,9,Divorced,Sales,Other-relative,White,Female,0,0,40,United-States,<=50K
7132,29,Private,129856,Assoc-voc,11,Never-married,Adm-clerical,Own-child,White,Female,0,0,40,United-States,<=50K
24194,47,Local-gov,126754,Masters,14,Married-civ-spouse,Prof-specialty,Husband,White,Male,0,0,40,United-States,>50K
2409,53,State-gov,50048,Doctorate,16,Married-civ-spouse,Prof-specialty,Husband,White,Male,7688,0,50,United-States,>50K
17757,26,Private,135645,Bachelors,13,Never-married,Sales,Not-in-family,White,Male,0,0,40,United-States,<=50K


In [7]:
df.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education.num     0
marital.status    0
occupation        0
relationship      0
race              0
sex               0
capital.gain      0
capital.loss      0
hours.per.week    0
native.country    0
income            0
dtype: int64

##### Task 1 — Fix Hidden Missing Values

In [8]:
df.replace('?', np.nan, inplace=True)
df.isnull().sum()

age                  0
workclass         1836
fnlwgt               0
education            0
education.num        0
marital.status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital.gain         0
capital.loss         0
hours.per.week       0
native.country     583
income               0
dtype: int64

##### Task 2 — Quantify Missingness

In [9]:
df.isnull().sum().sort_values(ascending=False)

occupation        1843
workclass         1836
native.country     583
fnlwgt               0
education            0
education.num        0
age                  0
marital.status       0
relationship         0
sex                  0
race                 0
capital.gain         0
capital.loss         0
hours.per.week       0
income               0
dtype: int64

In [10]:
(df.isnull().mean() * 100).sort_values(ascending=False)

occupation        5.660146
workclass         5.638647
native.country    1.790486
fnlwgt            0.000000
education         0.000000
education.num     0.000000
age               0.000000
marital.status    0.000000
relationship      0.000000
sex               0.000000
race              0.000000
capital.gain      0.000000
capital.loss      0.000000
hours.per.week    0.000000
income            0.000000
dtype: float64

##### Task 3 — Handle Missing Values

In [11]:
df['workclass'].value_counts()

workclass
Private             22696
Self-emp-not-inc     2541
Local-gov            2093
State-gov            1298
Self-emp-inc         1116
Federal-gov           960
Without-pay            14
Never-worked            7
Name: count, dtype: int64

In [12]:
df['occupation'].value_counts()

occupation
Prof-specialty       4140
Craft-repair         4099
Exec-managerial      4066
Adm-clerical         3770
Sales                3650
Other-service        3295
Machine-op-inspct    2002
Transport-moving     1597
Handlers-cleaners    1370
Farming-fishing       994
Tech-support          928
Protective-serv       649
Priv-house-serv       149
Armed-Forces            9
Name: count, dtype: int64

In [13]:
df['workclass'] = df['workclass'].fillna('Unknown')
df['occupation'] = df['occupation'].fillna('Unknown')
df['native.country'] = df['native.country'].fillna(df['native.country'].mode()[0])
df.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education.num     0
marital.status    0
occupation        0
relationship      0
race              0
sex               0
capital.gain      0
capital.loss      0
hours.per.week    0
native.country    0
income            0
dtype: int64

In [14]:
df[['age', 'fnlwgt', 'education.num', 'capital.gain', 'capital.loss', 'hours.per.week']].isnull().sum()

age               0
fnlwgt            0
education.num     0
capital.gain      0
capital.loss      0
hours.per.week    0
dtype: int64

**Justification:** `workclass` and `occupation` are imputed with an `'Unknown'` category instead of being dropped, since they make up a meaningful share of rows and dropping them would throw away usable data in every other column too. `native.country` is imputed with its mode (`United-States`) since missing values are a small share of a heavily one-sided column, so imputing barely changes the distribution. None of the numeric columns have missing values in this dataset once the `'?'` placeholders are fixed, so no numeric imputation is needed.

##### Task 4 — Detect Duplicates

In [15]:
df.duplicated().sum()

np.int64(24)

In [16]:
df[df.duplicated(keep=False)]

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
6227,90,Private,52386,Some-college,10,Never-married,Other-service,Not-in-family,Asian-Pac-Islander,Male,0,0,35,United-States,<=50K
7615,19,Private,251579,Some-college,10,Never-married,Other-service,Own-child,White,Male,0,0,14,United-States,<=50K
7978,25,Private,308144,Bachelors,13,Never-married,Craft-repair,Not-in-family,White,Male,0,0,40,Mexico,<=50K
8356,21,Private,250051,Some-college,10,Never-married,Prof-specialty,Own-child,White,Female,0,0,10,United-States,<=50K
8453,25,Private,308144,Bachelors,13,Never-married,Craft-repair,Not-in-family,White,Male,0,0,40,Mexico,<=50K
8500,38,Private,207202,HS-grad,9,Married-civ-spouse,Machine-op-inspct,Husband,White,Male,0,0,48,United-States,>50K
8645,90,Private,52386,Some-college,10,Never-married,Other-service,Not-in-family,Asian-Pac-Islander,Male,0,0,35,United-States,<=50K
9067,27,Private,255582,HS-grad,9,Never-married,Machine-op-inspct,Not-in-family,White,Female,0,0,40,United-States,<=50K
9269,20,Private,107658,Some-college,10,Never-married,Tech-support,Not-in-family,White,Female,0,0,10,United-States,<=50K
9305,25,Private,195994,1st-4th,2,Never-married,Priv-house-serv,Not-in-family,White,Female,0,0,40,Guatemala,<=50K


In [17]:
df.drop(columns=['income']).duplicated().sum()

np.int64(25)

Rows that are duplicates on every column except `income` are treated separately from exact duplicates: two different people can share identical demographic attributes yet land in different income brackets, so these are not necessarily data-entry errors. Only exact full-row duplicates are dropped below.

In [18]:
df = df.drop_duplicates()
df.shape

(32537, 15)

##### Task 5 — Fix Inconsistent Categorical Data

In [19]:
df['education'].unique()

array(['HS-grad', 'Some-college', '7th-8th', '10th', 'Doctorate',
       'Prof-school', 'Bachelors', 'Masters', '11th', 'Assoc-acdm',
       'Assoc-voc', '1st-4th', '5th-6th', '12th', '9th', 'Preschool'],
      dtype=object)

In [20]:
df['marital.status'].unique()

array(['Widowed', 'Divorced', 'Separated', 'Never-married',
       'Married-civ-spouse', 'Married-spouse-absent', 'Married-AF-spouse'],
      dtype=object)

In [21]:
df['native.country'].unique()

array(['United-States', 'Mexico', 'Greece', 'Vietnam', 'China', 'Taiwan',
       'India', 'Philippines', 'Trinadad&Tobago', 'Canada', 'South',
       'Holand-Netherlands', 'Puerto-Rico', 'Poland', 'Iran', 'England',
       'Germany', 'Italy', 'Japan', 'Hong', 'Honduras', 'Cuba', 'Ireland',
       'Cambodia', 'Peru', 'Nicaragua', 'Dominican-Republic', 'Haiti',
       'El-Salvador', 'Hungary', 'Columbia', 'Guatemala', 'Jamaica',
       'Ecuador', 'France', 'Yugoslavia', 'Scotland', 'Portugal', 'Laos',
       'Thailand', 'Outlying-US(Guam-USVI-etc)'], dtype=object)

In [22]:
cat_cols = ['workclass', 'education', 'marital.status', 'occupation', 'relationship', 'race', 'sex', 'native.country', 'income']
for col in cat_cols:
    df[col] = df[col].str.strip()
df['education'].unique()

array(['HS-grad', 'Some-college', '7th-8th', '10th', 'Doctorate',
       'Prof-school', 'Bachelors', 'Masters', '11th', 'Assoc-acdm',
       'Assoc-voc', '1st-4th', '5th-6th', '12th', '9th', 'Preschool'],
      dtype=object)

##### Task 6 — Descriptive Statistics

In [23]:
df.describe()

,age,fnlwgt,education.num,capital.gain,capital.loss,hours.per.week
count,32537.000000,3.253700e+04,32537.000000,32537.000000,32537.000000,32537.000000
mean,38.585549,1.897808e+05,10.081815,1078.443741,87.368227,40.440329
std,13.637984,1.055565e+05,2.571633,7387.957424,403.101833,12.346889
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.369930e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


In [24]:
df.describe(include='object')

,workclass,education,marital.status,occupation,relationship,race,sex,native.country,income
count,32537,32537,32537,32537,32537,32537,32537,32537,32537
unique,9,16,7,15,6,5,2,41,2
top,Private,HS-grad,Married-civ-spouse,Prof-specialty,Husband,White,Male,United-States,<=50K
freq,22673,10494,14970,4136,13187,27795,21775,29735,24698


In [25]:
age_mean, age_median = df['age'].mean(), df['age'].median()
hpw_mean, hpw_median = df['hours.per.week'].mean(), df['hours.per.week'].median()
age_mean, age_median, hpw_mean, hpw_median

(np.float64(38.585548759873376), 37.0, np.float64(40.44032947106371), 40.0)

Mean age (~38.6) and median age (37) are close, indicating a fairly symmetric age distribution with only mild right skew. Mean hours-per-week (~40.4) and median (40) are almost identical, showing that most people work close to a standard 40-hour week, with a smaller number of high-hour outliers pulling the mean slightly above the median.

##### Task 7 — Frequency Analysis

In [26]:
df['occupation'].value_counts()

occupation
Prof-specialty       4136
Craft-repair         4094
Exec-managerial      4065
Adm-clerical         3768
Sales                3650
Other-service        3291
Machine-op-inspct    2000
Unknown              1843
Transport-moving     1597
Handlers-cleaners    1369
Farming-fishing       992
Tech-support          927
Protective-serv       649
Priv-house-serv       147
Armed-Forces            9
Name: count, dtype: int64

In [27]:
df['sex'].value_counts(normalize=True) * 100

sex
Male      66.92381
Female    33.07619
Name: proportion, dtype: float64

In [28]:
df['income'].value_counts(normalize=True) * 100

income
<=50K    75.907428
>50K     24.092572
Name: proportion, dtype: float64

The most common occupation is `Prof-specialty` (closely followed by `Craft-repair` and `Exec-managerial`). The `sex` distribution is imbalanced (roughly two-thirds male, one-third female). The `income` classes are also imbalanced: roughly three-quarters of records fall in `<=50K` versus about a quarter in `>50K`, so this is an imbalanced classification target.

##### Task 8 — Cross-Check a Data Quality Assumption

In [29]:
edu_check = df.groupby('education')['education.num'].nunique()
edu_check

education
10th            1
11th            1
12th            1
1st-4th         1
5th-6th         1
7th-8th         1
9th             1
Assoc-acdm      1
Assoc-voc       1
Bachelors       1
Doctorate       1
HS-grad         1
Masters         1
Preschool       1
Prof-school     1
Some-college    1
Name: education.num, dtype: int64

In [30]:
edu_check[edu_check > 1]

Series([], Name: education.num, dtype: int64)

Each `education` label maps to exactly one `education.num` value with no inconsistencies found — the two columns are fully consistent encodings of the same information.

##### Task 9 — SQL-Based Data Extraction

In [31]:
import sqlite3

conn = sqlite3.connect('adult_income.db')
df.to_sql('adult_income', conn, if_exists='replace', index=False)

32537

In [32]:
query = "SELECT * FROM adult_income WHERE age > 30"
df_sql = pd.read_sql_query(query, conn)
df_sql.head()

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,Unknown,77053,HS-grad,9,Widowed,Unknown,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,Unknown,186061,Some-college,10,Widowed,Unknown,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


In [33]:
df_sql.shape, df[df['age'] > 30].shape

((21979, 15), (21979, 15))

In [34]:
conn.close()

##### Task 10 — Data Profiling Summary Report

###### Dataset Overview
The raw dataset contains 32,561 rows and 15 columns; after removing exact duplicates the cleaned dataset has **32,537 rows** and **15 columns**. It is the UCI Adult Income (Census Income) dataset — each row represents an individual from the 1994 US Census database, with demographic and employment attributes (age, workclass, education, occupation, marital status, hours worked per week, etc.) used to predict whether that person's annual income is above or below $50K.

###### Data Quality Issues Found
- **Missing values:** Hidden as the string `'?'` rather than true `NaN`, so `.isnull().sum()` initially reported zero missing values. After replacement: `occupation` had 1,843 missing, `workclass` had 1,836 missing, and `native.country` had 583 missing.
- **Duplicates:** 24 exact duplicate rows were found and removed. Checking duplicates while ignoring the `income` column found 25 matches (one additional row pair that only differs on `income`); these were kept rather than dropped, since they can plausibly represent different individuals with identical recorded attributes but different real outcomes.
- **Inconsistent data:** Categorical text columns (e.g., `education`, `marital.status`, `native.country`, `occupation`) contained leading/trailing whitespace, which was stripped.

###### Key Observations
- `income` classes are imbalanced: ~75.9% `<=50K` vs ~24.1% `>50K`, which matters for any downstream classification modeling (accuracy alone would be misleading).
- `sex` distribution is imbalanced: ~66.9% Male vs ~33.1% Female.
- The most common occupation is `Prof-specialty` (4,136 records), closely followed by `Craft-repair` and `Exec-managerial`.
- Mean age (~38.6) is close to median age (37); mean hours-per-week (~40.4) is close to median (40) — both distributions are roughly symmetric with only mild right skew from high-hour/older outliers.
- `education` and `education.num` are fully consistent — every education label maps to exactly one `education.num` value, with zero inconsistencies found.

###### Cleaning Decisions Made
- Replaced `'?'` placeholder strings with proper `NaN`.
- Imputed `workclass` and `occupation` missing values with an `'Unknown'` category rather than dropping rows, to retain the rest of each record's information instead of losing ~11% of rows.
- Imputed `native.country` missing values with its mode (`United-States`), since it is a small share (~1.8%) of a heavily skewed column.
- Dropped the 24 exact duplicate rows; retained rows that matched only on every column except the `income` label.
- Stripped whitespace from all categorical text columns to eliminate false "inconsistencies" in category values.